### DISCUSSION TITLE
Error in vCenter Scraper: Need to Specify Datacenter for Metric Collection

### DISCUSSION THREAD

**Discussion by sromlee** [open-telemetry/opentelemetry-collector-contrib#33654] [Q&A]
First of all, I'm a newbie and this is my very first time using Otel. The testing version is v0.102.0. The collector can produce some metrics for the datastore and host but no VM. 
Below is my error: 

`unable to retrieve ResourcePools with InventoryLists please specify a datacenter;`

Also the error about `querySpec`. Please see below is the full errors. 

`2024-06-19T12:19:50.542Z error scraperhelper/scrapercontroller.go:197 Error scraping metrics {"kind": "receiver", "name": "vcenter", "data_type": "metrics", "error": "unable to retrieve ResourcePools with InventoryLists: please specify a datacenter; unable to retrieve vApps with InventoryLists: please specify a datacenter; failed to retrieve perf metrics for HostSystems: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; failed to retrieve perf metrics for VirtualMachines: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected `

I'm not sure how to specify the data center or something went wrong with my configuration. 

Updated:
My VMware environment has no Resource pool/vApp/Cluster. How to avoid this kind of errors?

**Comment by sromlee** [comment]
Pinging code owners for receiver/vcenter: @djaglowski @schmikei @StefanKurek.

**Comment by StefanKurek** [comment]
@sromlee Hello. The user shouldn't have to manually specify a datacenter for collection and this should be picked up automatically. Are you connecting to vCenter or connecting directly to an ESXi host?

**Reply by sromlee** [reply]
@StefanKurek I’m connecting to the vCenter.

**Reply by StefanKurek** [reply]
@sromlee for the resources that you do get back, do they have a `vcenter.datacenter.name` or no?

**Reply by sromlee** [reply]
Yes, I’m able to get back the data center name. Some metrics that return are contain the attribute.

**Reply by StefanKurek** [reply]
@sromlee that's very interesting. I'm pretty confused about what might be going on here since you are able to see datacenter coming back. I'm doing some work right now updating the `govmomi` version for this receiver. If I find any insights while doing this work, I will get back with you.

**Reply by sromlee** [reply]
@StefanKurek Thank you for your reply.  I’m also confused. Seems vcenter.vm.* are not sent and it could not process the metrics due to no inventory path found.

**Reply by StefanKurek** [reply]
@sromlee did you say you had 1 VM metric that did return? If that's the case, can you post the resource attributes for that VM?

**Reply by sromlee** [reply]
@StefanKurek  I only get these resource attributes from metric : `vcenter.vm.disk.usage` I could not found any vm.name in resource_attributes as well. 
<img width="205" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/86556418/dccc7fbb-024b-4102-9412-60fa24ed23a2">

**Reply by StefanKurek** [reply]
@sromlee Can you tell me what version of vCenter you are using? I'm also curious if you only have a single datacenter in your environment (I don't think I already asked this)?

**Reply by sromlee** [reply]
@StefanKurek  I have multiple data centers with various VMs in each data center. 
Below is the detail of my Center.


```
Version: | 8.0.2
-- | --
Build: | 23929136
Clusters: | 0
Hosts: | 5
Virtual Machines: | 22

Version:	8.0.2
Build:	23929136
Clusters:	[0](javascript:void(0))
Hosts:	[5](javascript:void(0))
Virtual Machines:	[22](javascript:void(0))
```

**Comment by sromlee** [comment]
This is the full error. I'm not sure it's because of my environment is a standalone (not a cluster) ? 
@StefanKurek 
```
 error   scraperhelper/scrapercontroller.go:197  Error scraping metrics  {"kind": "receiver", "name": "vcenter", "data_type": "metrics", "error": "unable to retrieve ResourcePools with InventoryLists: please specify a datacenter; unable to retrieve vApps with InventoryLists: please specify a datacenter; failed to retrieve perf metrics for HostSystems: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; failed to retrieve perf metrics for VirtualMachines: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxxxxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [test]'s collected ResourcePool: Resources; no inventory path found for VM [test-vm]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [centos_template]'s collected ResourcePool: Resources; no inventory path found for VM [sg-vrep-6002]'s collected ResourcePool: Resources; no inventory path found for VM [centos_ide]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [sg-vrep-6001]'s collected ResourcePool: Resources; no inventory path found for VM [win2019]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources", "scraper": "vcenter"}
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport
        /go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.102.1/scraperhelper/scrapercontroller.go:197
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1
        /go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.102.1/scraperhelper/scrapercontroller.go:173
```

**Reply by StefanKurek** [reply]
@sromlee we have tested with standalone hosts, so that shouldn't be an issue.

**Comment by sromlee** [comment]
@djaglowski @schmikei @StefanKurek May I know if the scraper will work in case there is no resource pool, vapp and cluster ?
I noticed that the metric just sent only `vcenter.host` and `vcenter.datastore`. But for `vcenter.vm `just sent only 1 metric which is `vcenter.vm.disk.usage`.

**Reply by StefanKurek** [reply]
@sromlee bad news and good news. Bad news is you found a [bug](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33734) with the receiver. Some of the client calls were having issues with multiple datacenters present. The good news is I already have a [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33735) out to fix this. If you'd like, you can even build the collector with this branch to test the changes now, or you can wait until the next release. Thanks!

**Reply by sromlee** [reply]
Hi @StefanKurek  Thanks for your checking. I've tested your[PR]( https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33735) already and this time we see error:

```
2024-06-25T10:17:56.558+0700    error   scraperhelper/scrapercontroller.go:200  Error scraping metrics  {"kind": "receiver", "name": "vcenter", "data_type": "metrics", "error": "unable to get cluster lists: cluster '*' not found; unable to get cluster lists: cluster '*' not found; unable to get cluster lists: cluster '*' not found; unable to get cluster lists: cluster '*' not found", "errorCauses": [{"error": "unable to get cluster lists: cluster '*' not found"}, {"error": "unable to get cluster lists: cluster '*' not found"}, {"error": "unable to get cluster lists: cluster '*' not found"}, {"error": "unable to get cluster lists: cluster '*' not found"}], "scraper": "vcenter"}
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport
        go.opentelemetry.io/collector/receiver@v0.91.0/scraperhelper/scrapercontroller.go:200
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1
        go.opentelemetry.io/collector/receiver@v0.91.0/scraperhelper/scrapercontroller.go:176
```

Then it seems there is no metric return from the receiver.

**Reply by StefanKurek** [reply]
@sromlee are you sure that was from [this](https://github.com/observIQ/opentelemetry-collector-contrib/tree/fix/vcenter-multiple-datacenters) branch? That error message sounds like it is from a very outdated version of the receiver.

**Reply by sromlee** [reply]
@StefanKurek  Sorry for the confusion, yes the previous build was from the wrong branch. After correct the branch we still got the below error: 

This time it can get the data center properly but tried to get ResourcePool which does not exist (because we have no cluster)
The metrics return are the same with only 1 VM metric. 


```2024-06-25T12:26:41.506+0700    error   scraperhelper/scrapercontroller.go:197  Error scraping metrics  {"kind": "receiver", "name": "vcenter", "data_type": "metrics", "error": "unable to retrieve ResourcePools with InventoryLists for datacenter /xxx: resource pool '*' not found; failed to retrieve perf metrics for HostSystems: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; failed to retrieve perf metrics for VirtualMachines: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [2019]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [win2019_clean_install]'s collected ResourcePool: Resources; no inventory path found for collected ResourcePool: Resources; no inventory path found for VM [test-vm]'s collected ResourcePool: Resources; no inventory path found for VM [win2019]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [test]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [centos_template]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources; no inventory path found for VM [centos_ide]'s collected ResourcePool: Resources; no inventory path found for VM [xxxx]'s collected ResourcePool: Resources", "scraper": "vcenter"}```

**Reply by StefanKurek** [reply]
Hello again. Can you double check that the user being used at least has READ permissions for Resource Pools? Even though you are using standalone hosts, each should have a single default Resource Pool under the hood.

**Reply by sromlee** [reply]
@StefanKurek  My account has admin permission. And I see nothing in resource pool list. 

<img width="1357" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/86556418/52c5a68c-c61a-4cc0-978d-ea95dfb5a037">

**Reply by StefanKurek** [reply]
I wouldn't expect you to see it in the UI. From the VMware docs: 
<img width="1476" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/12396806/46535e3b-6c3e-4167-8a1d-cb25abad98b3">

**Reply by sromlee** [reply]
I just noticed from the error that. The error  `"unable to retrieve ResourcePools with InventoryLists for datacenter /xxx: resource pool '*' not found;` is happened to only 1 datacenter that has no host nor VM in it. It's just the empty datacenter. 

The rest datacenter has no such error. Is it possible that this is the root cause ?

**Reply by StefanKurek** [reply]
@sromlee that was most likely it. Pull in the latest changes and give that a try.

**Reply by sromlee** [reply]
@StefanKurek  The latest version is working! There was still an error for HostSystems. Btw, I can retrieve most of the metrics including all VMs metrics now.  Thank you very much for your help :) 

Just FYI, here is the error: 
`"error": "failed to retrieve perf metrics for HostSystems: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; failed to retrieve perf metrics for VirtualMachines: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0", "scraper": "vcenter"}`

**Reply by StefanKurek** [reply]
Is your data collection set to at least level 2? https://docs.vmware.com/en/VMware-vSphere/8.0/vsphere-monitoring-performance/GUID-25800DE4-68E5-41CC-82D9-8811E27924BC.html

**Reply by sromlee** [reply]
Yes, we set to level 2 already. Still facing the same error. 

```
2024-06-26T12:22:43.951547897+07:00 2024-06-26T12:22:43.951+0700 error scraperhelper/scrapercontroller.go:197 Error scraping metrics {"kind": "receiver", "name": "vcenter", "data_type": "metrics", "error": "failed to retrieve perf metrics for HostSystems: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; failed to retrieve perf metrics for VirtualMachines: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0", "scraper": "vcenter"}

2024-06-26T12:22:43.951627236+07:00 go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport

2024-06-26T12:22:43.951639027+07:00 go.opentelemetry.io/collector/receiver@v0.103.0/scraperhelper/scrapercontroller.go:197

2024-06-26T12:22:43.951646090+07:00 go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1

2024-06-26T12:22:43.951652943+07:00 go.opentelemetry.io/collector/receiver@v0.103.0/scraperhelper/scrapercontroller.go:173
```

<img width="1342" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/86556418/1f63dfc4-cb64-4724-a788-449f45531a34">

**Reply by StefanKurek** [reply]
@sromlee I don't think you should have to do this, but would you temporarily set it to level 4 to see if this makes a difference? I'm reaching at straws a bit here.

**Reply by sromlee** [reply]
@StefanKurek  I update to level 4 already but still the same error 

`2024-07-01T09:10:08.634711672+07:00 2024-07-01T09:10:08.634+0700 error scraperhelper/scrapercontroller.go:197 Error scraping metrics {"kind": "receiver", "name": "vcenter", "data_type": "metrics", "error": "failed to retrieve perf metrics for HostSystems: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0; failed to retrieve perf metrics for VirtualMachines: ServerFaultCode: \nRequired parameter querySpec is missing\n\nwhile parsing call information for method QueryPerf\nat line 2, column 66\n\nwhile parsing SOAP body\nat line 2, column 60\n\nwhile parsing SOAP envelope\nat line 2, column 0\n\nwhile parsing HTTP request for method queryStats\non object of type vim.PerformanceManager\nat line 1, column 0", "scraper": "vcenter"}`
